Fix cuda/cpu check on NoneType (Unit test)

mikekgfb opened the pull request:
Summary: Fix cuda/cpu check on NoneType (unit test)

Test Plan: sabdcastle/ github CI/CD

Differential Revision: D41208798

Skylion007 commented on torch/nn/modules/activation.py:
```diff
@@ -1113,7 +1113,7 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
             elif not all([(x is None or x.is_cuda or 'cpu' in str(x.device)) for x in tensor_args]):
                 why_not_fast_path = "some Tensor argument is neither CUDA nor CPU"
-            elif torch.is_grad_enabled() and any([x.requires_grad for x in tensor_args]):
+            elif torch.is_grad_enabled() and any([x is not None and x.requires_grad for x in tensor_args]):
```
```suggestion
            elif torch.is_grad_enabled() and any((x is not None and x.requires_grad for x in tensor_args))
```
is more efficient and allows for shortcircuiting (while the list comprehension version does not). This should probably be applied all over this file though.

Commit:
Fix cuda/cpu check on NoneType (Unit test)

Summary: Fix cuda/cpu check on NoneType (unit test)

Test Plan: sabdcastle/ github CI/CD

Differential Revision: D41208798

fbshipit-source-id: 3d55413ad8c834d5461be8627df91a76ad67cc9c

mikekgfb commented on torch/nn/modules/activation.py:
```diff
@@ -1113,7 +1113,7 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
                 why_not_fast_path = "some Tensor argument has_torch_function"
             elif not all([(x is None or x.is_cuda or 'cpu' in str(x.device)) for x in tensor_args]):
```
I think we need the list comprehension because of torchscriptability, as generators don't work in TS and we have a substantial number of users still depending on TS. :(

Skylion007 replied:
ooof, makes sense.

mikekgfb replied:
:(

mikekgfb commented:
@pytorchbot merge

mikekgfb commented:
@pytorchbot merge

malfet commented:
It would be nice to have a more detailed description on the nature of the fix and whether or not this is a regression (which is criteria for 1.13.1)

weiwangmeta commented:
Fixes: #88669
